src/components/infinite-scroll/InfiniteScroll.tsx

```diff
@@ -0,0 +1,73 @@
+import React, { useEffect, useRef, useState } from 'react'
+import styles from './InfiniteScroll.module.css'
+
+type InfiniteScrollProps<T> = {
+  getDataFunc: (page: number, offset: number) => Promise<T[]>
+  renderItem: (item: T) => JSX.Element
+  maxItemsInList: number
+  itemsPerPage: number
+}
+
+export const InfiniteScroll = <T,>({
+  getDataFunc,
+  renderItem,
+  maxItemsInList,
+  itemsPerPage,
+}: InfiniteScrollProps<T>) => {
+  const [items, setItems] = useState<T[]>([])
+  const [page, setPage] = useState(0)
+  const [isLoading, setIsLoading] = useState(false)
+
+  const observerTarget = useRef<HTMLDivElement>(null)
+
+  useEffect(() => {
+    if (
+      !observerTarget.current ||
+      isLoading ||
+      (maxItemsInList !== undefined && page * itemsPerPage >= maxItemsInList)
+    )
+      return
+
+    const fetchData = async () => {
+      setIsLoading(true)
+
+      const data = await getDataFunc(page, itemsPerPage)
```

Comment:
If getDataFunc fails, the component will always be stuck in the loading state. Consider using try/finally for error handling in this case